src/mlpack/methods/reinforcement_learning/q_networks/categorical_dqn.hpp

```diff
@@ -0,0 +1,201 @@
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_RL_CATEGORICAL_DQN_HPP
+#define MLPACK_METHODS_RL_CATEGORICAL_DQN_HPP
+
+#include <mlpack/prereqs.hpp>
+#include <mlpack/methods/ann/ffn.hpp>
+#include <mlpack/methods/ann/init_rules/gaussian_init.hpp>
+#include <mlpack/methods/ann/layer/layer.hpp>
+#include <mlpack/methods/ann/loss_functions/mean_squared_error.hpp>
+#include <mlpack/methods/ann/loss_functions/empty_loss.hpp>
+
+namespace mlpack {
+namespace rl {
+
+using namespace mlpack::ann;
+
+/**
+ * @tparam NetworkType The type of network used for categorical dqn.
+ */
+template <typename NetworkType = FFN<EmptyLoss<>, GaussianInitialization>>
+class CategoricalDQN
+{
+ public:
+  /**
+   * Default constructor.
+   */
+  CategoricalDQN() : network(), isNoisy(false), atomSize(0) 
+  { /* Nothing to do here. */ }
+
+  /**
+   * Construct an instance of CategoricalDQN class.
+   *
+   * @param inputDim Number of inputs.
+   * @param h1 Number of neurons in hiddenlayer-1.
+   * @param h2 Number of neurons in hiddenlayer-2.
+   * @param outputDim Number of neurons in output layer.
+   */
+  CategoricalDQN(const int inputDim,
+            const int h1,
+            const int h2,
+            const int outputDim,
+            const bool isNoisy = false,
+            const size_t atomSize = 51):
+      network(EmptyLoss<>(), GaussianInitialization(0, 0.001)),
+      isNoisy(isNoisy),
+      atomSize(atomSize)
+  {
+    network.Add(new Linear<>(inputDim, h1));
+    network.Add(new ReLULayer<>());
+    if(isNoisy)
+    { 
+      noisyLayerIndex.push_back(network.Model().size());
+      network.Add(new NoisyLinear<>(h1, h2));
+      network.Add(new ReLULayer<>());
+      noisyLayerIndex.push_back(network.Model().size());
+      network.Add(new NoisyLinear<>(h2, outputDim * atomSize));
+    }
+    else
+    {
+      network.Add(new Linear<>(h1, outputDim * atomSize));
+    }
+  }
+
+  CategoricalDQN(NetworkType network, const bool isNoisy, size_t atomSize):
+      network(std::move(network)),
+      isNoisy(isNoisy),
+      atomSize(atomSize)
+  { /* Nothing to do here. */ }
+
+  /**
+   * Predict the responses to a given set of predictors. The responses will
+   * reflect the output of the given output layer as returned by the
+   * output layer function.
+   *
+   * If you want to pass in a parameter and discard the original parameter
+   * object, be sure to use std::move to avoid unnecessary copy.
+   *
+   * @param state Input state.
+   * @param actionValue Matrix to put output action values of states input.
+   */
+  void Predict(const arma::mat state, arma::mat& actionValue)
+  {
+    arma::mat q_atoms;
+    std::cout << "network params:" <<  network.Parameters() << std::endl;
+    network.Predict(state, q_atoms);
+    std::cout << "q_atoms:" <<  q_atoms << std::endl;
```

Comment:
For comparision, I print the network parameters and its output. Since all weights and all biases are equal, shouldn't I expect an output of 4 values which are equal?

I do get 4 equal values, but after the first run, and not in the first run. Also, after the first run, using the debugger, I find that these 4 values are not **exactly** equal. Meaning, there is a difference in the range of `0.001`.

Reply:
I hope I make sense. Please let me know if I am unclear anywhere.

Reply:
I think I get what you mean, have you checked that in the linear layer the weight and bias are correct:

https://github.com/mlpack/mlpack/blob/60c928b0f8697f120b25ddc21e8b72fe52f8e8b0/src/mlpack/methods/ann/layer/linear_impl.hpp#L59-L60

In mlpack the bias comes after the weight, not sure this is the same for pytorch.

Reply:
In pytorch, weights and biases are not stored in a single vector/matrix. Rather they are segregated. But in mlpack, I'm pretty sure that 
```c++
  arma::mat params = {2.2145,2.2145,2.2145,2.2145,2.2145,2.2145,2.2145,2.2145,
                      -0.274,-0.274, 
                      2.2145,2.2145,2.2145,2.2145,2.2145,2.2145,2.2145,2.2145,
                      -0.274,-0.274, -0.274,-0.274};
```
should set values correctly. I even tried it separately, like here:
```c++
FFN<EmptyLoss<>, GaussianInitialization>
        qNetwork(EmptyLoss<>(), GaussianInitialization(0, 1));
    qNetwork.Add<Linear<>>(4, 2);
    qNetwork.Add(new ReLULayer<>());
    qNetwork.Add(new Linear<>(2, 4));
    qNetwork.ResetParameters();

    arma::colvec params = {2.2145, 2.2145, 2.2145, 2.2145, 2.2145, 2.2145, 2.2145, 2.2145,
                           -0.274, -0.274,
                           2.2145, 2.2145, 2.2145, 2.2145, 2.2145, 2.2145, 2.2145, 2.2145,
                           -0.274, -0.274, -0.274, -0.274};
    qNetwork.Parameters() = params;


    arma::colvec i = {1.2, 0.1, -5, 0.8};
    arma::colvec o;

    // arma::colvec b = {2, 0, -1, 0, 0, 0, 1, 0, 1, 10000, 2, 2, 10};
    // qNetwork.Parameters() = b;

    qNetwork.Predict(i, o);
    std::cout << o << std::endl;
```

the values are the same no matter what `i` I choose

Reply:
Hm, I tested the code above and I get a different value if I change `i`, for example `-5` to `-1` as well if I change `params`, so if they align (weight/bias) I think the model are correctly set. So it's strange that you don't get different values. Do you build with `DEBUG=ON`?

Reply:
Or maybe you just changed the values in a negative direction? The output of the rectifier function is 0 for all x < 0.

Reply:
I use g++ for all testing purposes, so i dont use cmake. And with `i` set to `{1.2, 0.1, -1, 0.8};`, and I get `o` as `9.3013` four times. Also, with all weights and biases equal, the output should be same right, irrespective of the input?

Reply:
I mean, the `o` output will contain equal numbers right?

Reply:
> Or maybe you just changed the values in a negative direction? The output of the rectifier function is 0 for all x < 0.

Ohh.. I see what u mean.. all I want to say is the output `o`  will be of the form `{x1, x2, x3, x4}`, where x1to4 will change according to input, but `x1=x2=x3=x4` holds true always.

Reply:
Here is the output of the network for each layer:

```
linear input:    1.2000   0.1000  -5.0000   0.8000

linear weight:    2.2145   2.2145
   2.2145   2.2145
   2.2145   2.2145
   2.2145   2.2145

linear bias:   -0.2740  -0.2740

linear output before bias:   -6.4221  -6.4221

linear output after bias:   -6.6961  -6.6961

rectifier input:   -6.6961  -6.6961

rectifier output:         0        0

linear input:         0        0

linear weight:    2.2145   2.2145   2.2145   2.2145
   2.2145   2.2145   2.2145   2.2145

linear bias:   -0.2740  -0.2740  -0.2740  -0.2740

linear output before bias:         0        0        0        0

linear output after bias:   -0.2740  -0.2740  -0.2740  -0.2740
```

With the same weights and bias for each layer do you expect a different result?

Reply:
No, i too am getting the same values, and the values are as expected. This is correct. But in the categorical implementation, if i set weights and biases equal, i don't see an exact `-0.2740` for all four values of `o`.

Reply:
So let's see if I get this right:

```c++
FFN<EmptyLoss<>, GaussianInitialization>
        qNetwork(EmptyLoss<>(), GaussianInitialization(0, 1));
    qNetwork.Add<Linear<>>(4, 2);
    qNetwork.Add(new ReLULayer<>());
    qNetwork.Add(new Linear<>(2, 4));
    qNetwork.ResetParameters();

    arma::colvec params = {2.2145, 2.2145, 2.2145, 2.2145, 2.2145, 2.2145, 2.2145, 2.2145,
                           -0.274, -0.274,
                           2.2145, 2.2145, 2.2145, 2.2145, 2.2145, 2.2145, 2.2145, 2.2145,
                           -0.274, -0.274, -0.274, -0.274};
    qNetwork.Parameters() = params;


    arma::colvec i = {1.2, 0.1, -5, 0.8};
    arma::colvec o;

    // arma::colvec b = {2, 0, -1, 0, 0, 0, 1, 0, 1, 10000, 2, 2, 10};
    // qNetwork.Parameters() = b;

    qNetwork.Predict(i, o);
    std::cout << o << std::endl;

```

produces the correct output, but 

```C++
  GreedyPolicy<CartPole> policy(1.0, 1000, 0.1, 1.0);
  RandomReplay<CartPole> replayMethod(1, 10000);

  TrainingConfig config;
  config.StepSize() = 0.001;
  config.Discount() = 0.99;
  config.TargetNetworkSyncInterval() = 1;
  config.ExplorationSteps() = 0;
  config.DoubleQLearning() = false;
  config.IsCategorical() = true;

  CategoricalDQN<> model(4, 2, 0, 2, false, 2);
```

does not?

Reply:
That is correct!

Reply:
Okay got it, can you post the complete test code you use?

Reply:
> Okay got it, can you post the complete test code you use?

Sure, here it is:
```c++
#include <mlpack/core.hpp>

#include "../mlpack/src/mlpack/methods/ann/ffn.hpp"
#include "../mlpack/src/mlpack/methods/ann/init_rules/gaussian_init.hpp"
#include "../mlpack/src/mlpack/methods/ann/layer/layer.hpp"
#include "../mlpack/src/mlpack/methods/ann/loss_functions/mean_squared_error.hpp"
#include "../mlpack/src/mlpack/methods/reinforcement_learning/environment/acrobot.hpp"
#include "../mlpack/src/mlpack/methods/reinforcement_learning/environment/mountain_car.hpp"
#include "../mlpack/src/mlpack/methods/reinforcement_learning/environment/cart_pole.hpp"
#include "../mlpack/src/mlpack/methods/reinforcement_learning/policy/greedy_policy.hpp"
#include "../mlpack/src/mlpack/methods/reinforcement_learning/q_learning.hpp"
#include "../mlpack/src/mlpack/methods/reinforcement_learning/q_networks/categorical_dqn.hpp"
#include "../mlpack/src/mlpack/methods/reinforcement_learning/training_config.hpp"

#include <ensmallen.hpp>
#include <typeinfo>

using namespace mlpack;
using namespace mlpack::ann;
using namespace ens;
using namespace mlpack::rl;

int main()
{
  mlpack::math::RandomSeed(std::time(NULL));
  // Set up the policy and replay method.
  GreedyPolicy<CartPole> policy(1.0, 1000, 0.1, 1.0);
  RandomReplay<CartPole> replayMethod(1, 10000);

  TrainingConfig config;
  config.StepSize() = 0.001;
  config.Discount() = 0.99;
  config.TargetNetworkSyncInterval() = 1;
  config.ExplorationSteps() = 0;
  config.DoubleQLearning() = false;
  config.IsCategorical() = true;

  CategoricalDQN<> model(4, 2, 0, 2, false, 2);

  // Set up DQN agent.
  QLearning<CartPole, decltype(model), AdamUpdate, decltype(policy)>
      agent(config, model, policy, replayMethod);

  arma::running_stat<double> averageReturn;
  size_t episodes = 0;
  bool converged = true;
  while (true)
  {
    double episodeReturn = agent.Episode();
    averageReturn(episodeReturn);
    episodes += 1;

    if (episodes > 1000)
    {
      std::cout << "Cart Pole with DQN failed." << std::endl;
      converged = false;
      break;
    }

    /**
     * Reaching running average return 35 is enough to show it works.
     * For the speed of the test case, I didn't set high criterion.
     */
    std::cout << "Average return: " << averageReturn.mean()
        << " Episode return: " << episodeReturn << std::endl;
    if (averageReturn.mean() > 150)
    {
      agent.Deterministic() = true;
      arma::running_stat<double> testReturn;
      for (size_t i = 0; i < 10; ++i)
        testReturn(agent.Episode());

      std::cout << "Average return in deterministic test: "
          << testReturn.mean() << std::endl;
      break;
    }
  }
}
```

Reply:
So turns out that the static weights are not set, at least they are not what is used in the linear layer.

Reply:
> So turns out that the static weights are not set, at least they are not what is used in the linear layer, at least not in the first run.

Exactly!

Reply:
So when I call `targetNetwork.ResetParameters();` right after we copy the learningNetwork:

https://github.com/mlpack/mlpack/blob/7c5198118cffebf6da2caa149e4c4d18886239ec/src/mlpack/methods/reinforcement_learning/q_learning_impl.hpp#L65

and store the static weights in an extra parameter:

```c++
arma::mat staticWeights = arma::mat(params);

learningNetwork.Parameters() = staticWeights;
targetNetwork.Parameters() = staticWeights;
```

I can see that the parameters are correct and the output as well.

Reply:
@nishantkr18 can you test the changes I mentioned above?

Reply:
Sorry for the delayed response, yes this seems to be working. I'll move ahead with the tests. Btw, why do you think that happened? What should be the correct way of copying over any network parameters?

Here, I see that if I `targetNetwork.ResetParameters();` after copying the network, the paramters become unequal. So I'd have to perform `targetNetwork.Parameters() = learningNetwork.Parameters();` just after reseting the parameters, right?

Reply:
Yes that is correct. The reason for this is that the copy constructor:
https://github.com/mlpack/mlpack/blob/f500227c6492c8cdcabd407b3bd028565c99f218/src/mlpack/methods/ann/ffn_impl.hpp#L644-L651

does not copy the weights, so we should call `ResetParameters()` first and assign the weights as you mentioned in the copy constructor itself.